src/argilla/training/base.py

```diff
@@ -0,0 +1,163 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import inspect
+import logging
+from typing import List, Union
+
+import argilla as rg
+from argilla.training.setfit import ArgillaSetFitTrainer
+
+
+class ArgillaBaseTrainer(object):
+    _logger = logging.getLogger("argilla.training")
+
+    def __init__(self, name: str, framework: str, train_size: float = None, seed: int = None, **load_kwargs):
+        """
+        `__init__` is a function that initializes the class
+
+        Args:
+          name (str): the name of the dataset you want to load.
+          framework (str): the framework to use for training. Currently, only "setfit" is supported.
+          query (str): a query to filter the dataset.
+          train_size (float): the size of the training set. If not specified, the entire dataset will be
+        used for training.
+          seed (int): int = None,
+        """
+        self.device = "cpu"
+
+        self._name = name
+        self._multi_label = False
+        self._split_applied = False
+        self._train_size = train_size
+        self._seed = seed
+
+        if train_size:
+            self._split_applied = True
+
+        self.rg_dataset_snapshot = rg.load(name=self._name, limit=1)
+        assert len(self.rg_dataset_snapshot) > 0, "Dataset must have at least one Validated record"
+        if isinstance(self.rg_dataset_snapshot, rg.DatasetForTextClassification):
+            self._rg_dataset_type = rg.DatasetForTextClassification
+            self._required_fields = ["id", "text", "inputs", "annotation", "multi_label"]
+            if self.rg_dataset_snapshot[0].multi_label:
+                self._multi_label = True
+        elif isinstance(self.rg_dataset_snapshot, rg.DatasetForTokenClassification):
+            self._rg_dataset_type = rg.DatasetForTokenClassification
+            self._required_fields = ["id", "text", "tokens", "ner_tags"]
+
+        elif isinstance(self.rg_dataset_snapshot, rg.DatasetForText2Text):
+            self._rg_dataset_type = rg.DatasetForText2Text
+            self._required_fields = ["id", "text", "annotation"]
+        else:
+            raise NotImplementedError(f"Dataset type {type(self.rg_dataset_snapshot)} is not supported.")
+
+        self.dataset_full = rg.load(name=self._name, fields=self._required_fields, **load_kwargs)
+        self.dataset_full_prepared = self.dataset_full.prepare_for_training(
+            framework=framework, train_size=self._train_size, seed=self._seed
+        )
+        if framework in ["transformers", "setfit"]:
+            import torch
+
+            if torch.backends.mps.is_available():
+                self.device = "mps"
+            elif torch.cuda.is_available():
+                self.device = "cuda"
+            else:
+                self.device = "cpu"
```

Comment:
Perhaps we can defer this logic to the more specific subclasses (`ArgillaSetFitTrainer`, `ArgillaTransformersTrainer`, etc.) instead? That way the base trainer needs to know less about the subclasses.

Reply:
I wanted to avoid copying this logic at least twice  (transformers, setfit). Maybe also spacy[transformers]. But I agree and moved it to the SetFit custom logic for now.